diff -Naur --strip-trailing-cr -x cclang -x .git -x .gitignore a/bindings/python/clang/cindex.py b/bindings/python/clang/cindex.py
--- a/bindings/python/clang/cindex.py	2017-10-17 14:06:40.160270346 +0300
+++ b/bindings/python/clang/cindex.py	2017-10-17 14:19:20.020308189 +0300
@@ -1887,6 +1887,7 @@
 TypeKind.OBJCCLASS = TypeKind(28)
 TypeKind.OBJCSEL = TypeKind(29)
 TypeKind.FLOAT128 = TypeKind(30)
+TypeKind.HALF = TypeKind(31)
 TypeKind.COMPLEX = TypeKind(100)
 TypeKind.POINTER = TypeKind(101)
 TypeKind.BLOCKPOINTER = TypeKind(102)
diff -Naur --strip-trailing-cr -x cclang -x .git -x .gitignore a/include/clang-c/Index.h b/include/clang-c/Index.h
--- a/include/clang-c/Index.h	2017-10-17 14:06:40.164270346 +0300
+++ b/include/clang-c/Index.h	2017-10-17 14:19:20.020308189 +0300
@@ -3011,8 +3011,9 @@
   CXType_ObjCClass = 28,
   CXType_ObjCSel = 29,
   CXType_Float128 = 30,
+  CXType_Half = 31,
   CXType_FirstBuiltin = CXType_Void,
-  CXType_LastBuiltin  = CXType_ObjCSel,
+  CXType_LastBuiltin  = CXType_Half,
 
   CXType_Complex = 100,
   CXType_Pointer = 101,
diff -Naur --strip-trailing-cr -x cclang -x .git -x .gitignore a/test/Index/opencl-types.cl b/test/Index/opencl-types.cl
--- a/test/Index/opencl-types.cl	1970-01-01 03:00:00.000000000 +0300
+++ b/test/Index/opencl-types.cl	2017-10-17 14:19:20.040308190 +0300
@@ -0,0 +1,24 @@
+// RUN: c-index-test -test-print-type %s | FileCheck %s
+
+#pragma OPENCL EXTENSION cl_khr_fp16 : enable
+#pragma OPENCL EXTENSION cl_khr_fp64 : enable
+
+typedef half half4 __attribute__((ext_vector_type(4)));
+typedef float float4 __attribute__((ext_vector_type(4)));
+typedef double double4 __attribute__((ext_vector_type(4)));
+
+void kernel testFloatTypes() {
+  half scalarHalf;
+  half4 vectorHalf;
+  float scalarFloat;
+  float4 vectorFloat;
+  double scalarDouble;
+  double4 vectorDouble;
+}
+
+// CHECK: VarDecl=scalarHalf:11:8 (Definition) [type=half] [typekind=Half] [isPOD=1]
+// CHECK: VarDecl=vectorHalf:12:9 (Definition) [type=half4] [typekind=Typedef] [canonicaltype=half __attribute__((ext_vector_type(4)))] [canonicaltypekind=Unexposed] [isPOD=1]
+// CHECK: VarDecl=scalarFloat:13:9 (Definition) [type=float] [typekind=Float] [isPOD=1]
+// CHECK: VarDecl=vectorFloat:14:10 (Definition) [type=float4] [typekind=Typedef] [canonicaltype=float __attribute__((ext_vector_type(4)))] [canonicaltypekind=Unexposed] [isPOD=1]
+// CHECK: VarDecl=scalarDouble:15:10 (Definition) [type=double] [typekind=Double] [isPOD=1]
+// CHECK: VarDecl=vectorDouble:16:11 (Definition) [type=double4] [typekind=Typedef] [canonicaltype=double __attribute__((ext_vector_type(4)))] [canonicaltypekind=Unexposed] [isPOD=1]
diff -Naur --strip-trailing-cr -x cclang -x .git -x .gitignore a/tools/libclang/CXType.cpp b/tools/libclang/CXType.cpp
--- a/tools/libclang/CXType.cpp	2017-10-17 14:06:40.328270354 +0300
+++ b/tools/libclang/CXType.cpp	2017-10-17 14:19:20.040308190 +0300
@@ -48,6 +48,7 @@
     BTCASE(Long);
     BTCASE(LongLong);
     BTCASE(Int128);
+    BTCASE(Half);
     BTCASE(Float);
     BTCASE(Double);
     BTCASE(LongDouble);
@@ -502,6 +503,7 @@
     TKIND(Long);
     TKIND(LongLong);
     TKIND(Int128);
+    TKIND(Half);
     TKIND(Float);
     TKIND(Double);
     TKIND(LongDouble);
