diff -Naur --strip-trailing-cr -x cclang -x .git -x .gitignore a/include/clang/Basic/Specifiers.h b/include/clang/Basic/Specifiers.h
--- a/include/clang/Basic/Specifiers.h	2017-10-17 14:06:40.184270347 +0300
+++ b/include/clang/Basic/Specifiers.h	2017-10-17 14:31:25.036344297 +0300
@@ -82,11 +82,12 @@
   /// \brief Structure that packs information about the type specifiers that
   /// were written in a particular type specifier sequence.
   struct WrittenBuiltinSpecs {
-    /*DeclSpec::TST*/ unsigned Type  : 5;
+    static_assert(TST_error < 1 << 6, "Type bitfield not wide enough for TST");
+    /*DeclSpec::TST*/ unsigned Type  : 6;
     /*DeclSpec::TSS*/ unsigned Sign  : 2;
     /*DeclSpec::TSW*/ unsigned Width : 2;
     unsigned ModeAttr : 1;
-  };  
+  };
 
   /// \brief A C++ access specifier (public, private, protected), plus the
   /// special value "none" which means different things in different contexts.
