diff -Naur --strip-trailing-cr -x cclang -x .git -x .gitignore a/lib/Frontend/CompilerInvocation.cpp b/lib/Frontend/CompilerInvocation.cpp
--- a/lib/Frontend/CompilerInvocation.cpp	2017-10-17 14:06:40.256270351 +0300
+++ b/lib/Frontend/CompilerInvocation.cpp	2017-10-17 15:01:23.820433882 +0300
@@ -1616,7 +1616,6 @@
   if (Opts.OpenCL) {
     Opts.AltiVec = 0;
     Opts.ZVector = 0;
-    Opts.CXXOperatorNames = 1;
     Opts.LaxVectorConversions = 0;
     Opts.DefaultFPContract = 1;
     Opts.NativeHalfType = 1;
