diff -Naur --strip-trailing-cr -x cclang -x .git -x .gitignore a/lib/Sema/SemaDecl.cpp b/lib/Sema/SemaDecl.cpp
--- a/lib/Sema/SemaDecl.cpp	2017-10-17 14:06:40.284270352 +0300
+++ b/lib/Sema/SemaDecl.cpp	2017-10-17 15:20:29.640490947 +0300
@@ -2101,7 +2101,9 @@
   // -Wtypedef-redefinition.  If either the original or the redefinition is
   // in a system header, don't emit this for compatibility with GCC.
   if (getDiagnostics().getSuppressSystemWarnings() &&
-      (Context.getSourceManager().isInSystemHeader(Old->getLocation()) ||
+      // Some standard types are defined implicitly in Clang (e.g. OpenCL).
+      (Old->isImplicit() ||
+       Context.getSourceManager().isInSystemHeader(Old->getLocation()) ||
        Context.getSourceManager().isInSystemHeader(New->getLocation())))
     return;
 
diff -Naur --strip-trailing-cr -x cclang -x .git -x .gitignore a/test/SemaOpenCL/types.cl b/test/SemaOpenCL/types.cl
--- a/test/SemaOpenCL/types.cl	1970-01-01 03:00:00.000000000 +0300
+++ b/test/SemaOpenCL/types.cl	2017-10-17 15:20:29.640490947 +0300
@@ -0,0 +1,6 @@
+// RUN: %clang_cc1 %s -cl-std=CL2.0 -verify -fsyntax-only
+
+// expected-no-diagnostics
+
+// Check redefinition of standard types
+typedef atomic_int atomic_flag;
