diff -Naur --strip-trailing-cr -x cclang -x .git -x .gitignore a/lib/Sema/SemaInit.cpp b/lib/Sema/SemaInit.cpp
--- a/lib/Sema/SemaInit.cpp	2017-10-17 15:34:19.236532264 +0300
+++ b/lib/Sema/SemaInit.cpp	2017-10-17 15:58:44.268605227 +0300
@@ -1203,7 +1203,7 @@
 
   } else {
     assert((ElemType->isRecordType() || ElemType->isVectorType() ||
-            ElemType->isClkEventT()) && "Unexpected type");
+            ElemType->isOpenCLSpecificType()) && "Unexpected type");
 
     // C99 6.7.8p13:
     //
diff -Naur --strip-trailing-cr -x cclang -x .git -x .gitignore a/test/SemaOpenCL/array-init.cl b/test/SemaOpenCL/array-init.cl
--- a/test/SemaOpenCL/array-init.cl	1970-01-01 03:00:00.000000000 +0300
+++ b/test/SemaOpenCL/array-init.cl	2017-10-17 15:58:44.268605227 +0300
@@ -0,0 +1,20 @@
+// RUN: %clang_cc1 %s -verify -pedantic -fsyntax-only -cl-std=CL2.0
+// expected-no-diagnostics
+
+__kernel void k1(queue_t q1, queue_t q2) {
+  queue_t q[] = {q1, q2};
+}
+
+__kernel void k2(read_only pipe int p) {
+  reserve_id_t i1 = reserve_read_pipe(p, 1);
+  reserve_id_t i2 = reserve_read_pipe(p, 1);
+  reserve_id_t i[] = {i1, i2};
+}
+
+event_t create_event();
+__kernel void k3() {
+  event_t e1 = create_event();
+  event_t e2 = create_event();
+  event_t e[] = {e1, e2};
+}
+
