diff -Naur --strip-trailing-cr -x cclang -x .git -x .gitignore a/test/SemaOpenCL/vector_literals_invalid.cl b/test/SemaOpenCL/vector_literals_invalid.cl
--- a/test/SemaOpenCL/vector_literals_invalid.cl	2017-10-17 14:06:40.328270354 +0300
+++ b/test/SemaOpenCL/vector_literals_invalid.cl	2017-10-17 16:36:01.256716635 +0300
@@ -1,4 +1,4 @@
-// RUN: %clang_cc1 -verify %s 
+// RUN: %clang_cc1 -verify %s
 
 typedef __attribute__(( ext_vector_type(4) ))  float float4;
 typedef __attribute__(( ext_vector_type(4) ))  int int4;
@@ -10,4 +10,5 @@
   int4 b = (int4)(1,2,3,4,5); // expected-error{{excess elements in vector}}
   ((float4)(1.0f))++; // expected-error{{cannot increment value of type 'float4'}}
   int8 d = (int8)(a,(float4)(1)); // expected-error{{initializing 'int' with an expression of incompatible type 'float4'}}
+  ((int4)(0)).x = 8; // expected-error{{expression is not assignable}}
 }
