diff -Naur --strip-trailing-cr -x cclang -x .git -x .gitignore a/lib/Sema/SemaExpr.cpp b/lib/Sema/SemaExpr.cpp
--- a/lib/Sema/SemaExpr.cpp	2017-10-17 19:43:39.233277317 +0300
+++ b/lib/Sema/SemaExpr.cpp	2017-10-17 20:07:35.993348872 +0300
@@ -13851,13 +13851,16 @@
   bool ByRef = false;
       
   // Blocks are not allowed to capture arrays.
-  if (CaptureType->isArrayType()) {
-    if (BuildAndDiagnose) {
-      S.Diag(Loc, diag::err_ref_array_type);
-      S.Diag(Var->getLocation(), diag::note_previous_decl) 
-      << Var->getDeclName();
+  // Only if it's not OpenCL 2.0.
+  if (!(S.getLangOpts().OpenCL && S.getLangOpts().OpenCLVersion >= 200)) {
+    if (CaptureType->isArrayType()) {
+      if (BuildAndDiagnose) {
+        S.Diag(Loc, diag::err_ref_array_type);
+        S.Diag(Var->getLocation(), diag::note_previous_decl)
+          << Var->getDeclName();
+      }
+      return false;
     }
-    return false;
   }
 
   // Forbid the block-capture of autoreleasing variables.
@@ -13889,7 +13892,11 @@
 
   const bool HasBlocksAttr = Var->hasAttr<BlocksAttr>();
   if (HasBlocksAttr || CaptureType->isReferenceType() ||
-      (S.getLangOpts().OpenMP && S.IsOpenMPCapturedDecl(Var))) {
+      (S.getLangOpts().OpenMP && S.IsOpenMPCapturedDecl(Var)) ||
+       // This is a unique behavior for OpenCL 2.0, since array capturing is
+       // allowed.
+      (S.getLangOpts().OpenCL && S.getLangOpts().OpenCLVersion >= 200 &&
+       CaptureType->isArrayType())) {
     // Block capture by reference does not change the capture or
     // declaration reference types.
     ByRef = true;
