From 9e78c38a7beb5044ac11785659d608cf31dda368 Mon Sep 17 00:00:00 2001
From: Feng Zou <feng.zou@intel.com>
Date: Tue, 20 Oct 2020 12:20:53 +0800
Subject: [PATCH] Remove repo name in LLVM IR

---
 cmake/modules/VersionFromVCS.cmake | 37 ++++++++++++++++--------------
 1 file changed, 20 insertions(+), 17 deletions(-)

diff --git a/cmake/modules/VersionFromVCS.cmake b/cmake/modules/VersionFromVCS.cmake
index 56331a3a81f..6d648df0ab6 100644
--- a/cmake/modules/VersionFromVCS.cmake
+++ b/cmake/modules/VersionFromVCS.cmake
@@ -17,9 +17,10 @@ function(get_source_info_svn path revision repository)
     if(Project_WC_REVISION)
       set(${revision} ${Project_WC_REVISION} PARENT_SCOPE)
     endif()
-    if(Project_WC_URL)
-      set(${repository} ${Project_WC_URL} PARENT_SCOPE)
-    endif()
+    # Do not show repo name in IR
+    # if(Project_WC_URL)
+    #   set(${repository} ${Project_WC_URL} PARENT_SCOPE)
+    # endif()
   endif()
 endfunction()
 
@@ -43,9 +44,10 @@ function(get_source_info_git path revision repository)
           string(REGEX REPLACE "^(.*\n)?Revision: ([^\n]+).*"
             "\\2" git_svn_rev "${git_output}")
           set(${revision} ${git_svn_rev} PARENT_SCOPE)
-          string(REGEX REPLACE "^(.*\n)?URL: ([^\n]+).*"
-            "\\2" git_url "${git_output}")
-          set(${repository} ${git_url} PARENT_SCOPE)
+          # Do not show repo name in IR
+          # string(REGEX REPLACE "^(.*\n)?URL: ([^\n]+).*"
+          #   "\\2" git_url "${git_output}")
+          # set(${repository} ${git_url} PARENT_SCOPE)
         endif()
       else()
         execute_process(COMMAND ${GIT_EXECUTABLE} rev-parse HEAD
@@ -67,17 +69,18 @@ function(get_source_info_git path revision repository)
         else()
           set(remote "origin")
         endif()
-        execute_process(COMMAND ${GIT_EXECUTABLE} remote get-url ${remote}
-          WORKING_DIRECTORY ${path}
-          RESULT_VARIABLE git_result
-          OUTPUT_VARIABLE git_output
-          ERROR_QUIET)
-        if(git_result EQUAL 0)
-          string(STRIP "${git_output}" git_output)
-          set(${repository} ${git_output} PARENT_SCOPE)
-        else()
-          set(${repository} ${path} PARENT_SCOPE)
-        endif()
+        # Do not show repo name in IR
+        # execute_process(COMMAND ${GIT_EXECUTABLE} remote get-url ${remote}
+        #   WORKING_DIRECTORY ${path}
+        #   RESULT_VARIABLE git_result
+        #   OUTPUT_VARIABLE git_output
+        #   ERROR_QUIET)
+        # if(git_result EQUAL 0)
+        #   string(STRIP "${git_output}" git_output)
+        #   set(${repository} ${git_output} PARENT_SCOPE)
+        # else()
+        #   set(${repository} ${path} PARENT_SCOPE)
+        # endif()
       endif()
     endif()
   endif()
-- 
2.18.1

